pkg/authorization/registry/policybinding/rest.go

```diff
@@ -4,96 +4,104 @@ import (
 	"fmt"
 
 	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
-	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/generic"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
-	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
 
 	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
 	"github.com/openshift/origin/pkg/authorization/api/validation"
 )
 
-// REST implements the RESTStorage interface in terms of an Registry.
-type REST struct {
-	registry Registry
+// strategy implements behavior for nodes
+type strategy struct {
+	runtime.ObjectTyper
 }
 
-// NewREST creates a new REST for policyBindings.
-func NewREST(registry Registry) *REST {
-	return &REST{registry}
+// Strategy is the default logic that applies when creating and updating PolicyBinding objects.
+var Strategy = strategy{kapi.Scheme}
+
+// NamespaceScoped is false for policies.
+func (strategy) NamespaceScoped() bool {
+	return true
 }
 
-// New creates a new PolicyBinding object
-func (r *REST) New() runtime.Object {
-	return &authorizationapi.PolicyBinding{}
+// AllowCreateOnUpdate is false for policies.
+func (strategy) AllowCreateOnUpdate() bool {
+	return false
 }
 
-func (r *REST) NewList() runtime.Object {
-	return &authorizationapi.PolicyBindingList{}
+func (strategy) GenerateName(base string) string {
+	return base
 }
 
-// List obtains a list of PolicyBindings that match selector.
-func (r *REST) List(ctx kapi.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
-	policyBindings, err := r.registry.ListPolicyBindings(ctx, label, field)
-	if err != nil {
-		return nil, err
-	}
-	return policyBindings, err
+// PrepareForCreate clears fields that are not allowed to be set by end users on creation.
+func (strategy) PrepareForCreate(obj runtime.Object) {
+	binding := obj.(*authorizationapi.PolicyBinding)
 
+	binding.Name = binding.PolicyRef.Namespace
+	scrubBindingRefs(binding)
 }
 
-// Get obtains the policyBinding specified by its id.
-func (r *REST) Get(ctx kapi.Context, id string) (runtime.Object, error) {
-	policyBinding, err := r.registry.GetPolicyBinding(ctx, id)
-	if err != nil {
-		return nil, err
+func scrubBindingRefs(binding *authorizationapi.PolicyBinding) {
+	binding.PolicyRef = kapi.ObjectReference{Namespace: binding.PolicyRef.Namespace}
+
+	for roleBindingKey, roleBinding := range binding.RoleBindings {
+		roleBinding.RoleRef = kapi.ObjectReference{Namespace: binding.PolicyRef.Namespace, Name: roleBinding.RoleRef.Name}
+		binding.RoleBindings[roleBindingKey] = roleBinding
```

Comment:
Shockingly well actually, although I'm careful not to change the keys and force a resize.  I'm willing to do it differently if you find it jarring.  It does kind of leap off the page.